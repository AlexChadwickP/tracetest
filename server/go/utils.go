package openapi

import (
	"net/http"

	"github.com/gogo/protobuf/jsonpb"
	"github.com/gogo/protobuf/proto"
	v11 "go.opentelemetry.io/proto/otlp/common/v1"
	res "go.opentelemetry.io/proto/otlp/resource/v1"
	v1 "go.opentelemetry.io/proto/otlp/trace/v1"
)

func EncodeJSONPBResponse(i interface{}, status *int, w http.ResponseWriter) error {
	w.Header().Set("Content-Type", "application/json; charset=UTF-8")
	if status != nil {
		w.WriteHeader(*status)
	} else {
		w.WriteHeader(http.StatusOK)
	}

	trace := i.(proto.Message)
	m := jsonpb.Marshaler{}
	return m.Marshal(w, trace)
}

func FixParent(tr *v1.TracesData, traceID, parentSpanID string) *v1.TracesData {
	spans := make(map[string]*v1.Span)
	for _, rs := range tr.ResourceSpans {
		for _, ils := range rs.InstrumentationLibrarySpans {
			for _, sp := range ils.Spans {
				spans[string(sp.SpanId)] = sp
			}
		}
	}

	// Fix parent id
	for _, sp := range spans {
		if sp.ParentSpanId == nil {
			continue
		}
		_, ok := spans[string(sp.ParentSpanId)]
		if !ok {
			sp.ParentSpanId = []byte(parentSpanID)
		}
	}
	rs := &v1.ResourceSpans{
		Resource: &res.Resource{
			Attributes:             []*v11.KeyValue{},
			DroppedAttributesCount: 0,
		},
		InstrumentationLibrarySpans: []*v1.InstrumentationLibrarySpans{
			{
				InstrumentationLibrary: &v11.InstrumentationLibrary{
					Name:    "tracetest",
					Version: "v1",
				},
				Spans: []*v1.Span{
					{TraceId: []byte(traceID),
						SpanId:       []byte(parentSpanID),
						ParentSpanId: nil,
						Name:         "tracetest",
						Kind:         v1.Span_SPAN_KIND_CLIENT,
					},
				},
			},
		},
		SchemaUrl: "",
	}
	tr.ResourceSpans = append(tr.ResourceSpans, rs)

	return tr
}
