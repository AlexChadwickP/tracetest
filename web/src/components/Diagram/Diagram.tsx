import {TSpan} from '../../types/Span.types';
import {TTrace} from '../../types/Trace.types';
import DAGComponent from './components/DAG';
import {TimelineChart} from './components/TimelineChart';

export enum SupportedDiagrams {
  DAG = 'dag',
  Timeline = 'timeline',
}

export interface IDiagramProps {
  type: SupportedDiagrams;
  trace: TTrace;
  selectedSpan?: TSpan;
  onSelectSpan?(spanId: string): void;
}

const ComponentMap: Record<string, typeof DAGComponent | typeof TimelineChart> = {
  [SupportedDiagrams.DAG]: DAGComponent,
  [SupportedDiagrams.Timeline]: TimelineChart,
};

const Diagram: React.FC<IDiagramProps> = ({type, ...props}) => {
  const Component = ComponentMap[type || ''] || DAGComponent;

  return <Component type={type} {...props} />;
};

export default Diagram;
