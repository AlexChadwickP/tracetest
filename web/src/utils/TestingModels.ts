export const TestingModels = {
  assertionResult: {
    assertion: {assertionId: '', selectors: [], spanAssertions: []},
    spanListAssertionResult: [],
  },
};
