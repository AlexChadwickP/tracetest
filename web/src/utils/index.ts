const flattenAttributesSelector = () =>
  `resourceSpans[].[instrumentationLibrarySpans[].spans[].attributes[].{key:key,value:value.*|[0]},resource.attributes[].{key:key,value: value.*|[0]}]|[][]`;

export const filterBySpanId = (spanId: string = '') =>
  `resourceSpans[?instrumentationLibrarySpans[?spans[?starts_with(spanId,'${spanId}')]]] | [].[instrumentationLibrarySpans[].spans[].attributes[].{key:key,value:value.*|[0],type:'span'},resource.attributes[].{key:key,value: value.*|[0],type:'resource'}]|[][]`;
