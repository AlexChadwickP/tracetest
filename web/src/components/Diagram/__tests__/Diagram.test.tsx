import {render} from '@testing-library/react';
import {ReactFlowProvider} from 'react-flow-renderer';
import {TestingModels} from '../../../utils/TestingModels';
import AnalyticsProvider from '../../Analytics';
import Diagram, {SupportedDiagrams} from '../Diagram';

test('Diagram DAG', () => {
  const result = render(
    <AnalyticsProvider>
      <ReactFlowProvider>
        <Diagram
          type={SupportedDiagrams.DAG}
          trace={TestingModels.trace}
          onSelectSpan={jest.fn()}
          selectedSpan={TestingModels.span}
        />
      </ReactFlowProvider>
    </AnalyticsProvider>
  );
  expect(result.container).toMatchSnapshot();
});
test('Diagram Timeline', () => {
  const result = render(
    <AnalyticsProvider>
      <ReactFlowProvider>
        <Diagram
          type={SupportedDiagrams.Timeline}
          trace={TestingModels.trace}
          onSelectSpan={jest.fn()}
          selectedSpan={TestingModels.span}
        />
      </ReactFlowProvider>
    </AnalyticsProvider>
  );
  expect(result.container).toMatchSnapshot();
});
