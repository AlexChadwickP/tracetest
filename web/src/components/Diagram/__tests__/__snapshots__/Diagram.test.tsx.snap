// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Diagram DAG 1`] = `
<div>
  <div
    data-cy="not_initialized"
  />
</div>
`;

exports[`Diagram Timeline 1`] = `
<div>
  <div
    data-cy="not_initialized"
  />
</div>
`;
